tenpy/algorithms/mps_compress.py

```diff
@@ -0,0 +1,345 @@
+        B=psi.get_B(0,form=None)
+        for i in range(psi.L - 1):
+            B=B.combine_legs(['vL', 'p'])
+            q,r =npc.qr(B, inner_labels=['vR', 'vL'])
+            B=q.split_legs()
+            psi.set_B(i,B,form=None)
+            B=psi.get_B((i+1)%L,form=None)
+            B=npc.tensordot(r,B, axes=('vR', 'vL'))
+        # Do SVD from right to left, truncate the singular values according to trunc_par
+        for i in range(psi.L-1,0,-1):
+            B=B.combine_legs(['p', 'vR'])
+            u, s, vh, err, norm_new = svd_theta(B, trunc_par)
+            psi.norm*=norm_new
+            vh=vh.split_legs()
+            psi.set_B(i%L, vh, form='B')
+            B=psi.get_B(i-1, form=None)
+            B=npc.tensordot(B, u, axes=('vR', 'vL'))
+            B.iscale_axis(s, 'vR')
+            psi.set_SL(i%L, s)
+        psi.set_B(0, B, form='Th')
+    if bc=='infinite':
+        for i in range(psi.L):
+            svd_two_site(i, psi)
+        for i in range(psi.L-1,-1, -1):
+            svd_two_site(i, psi, trunc_par)
+
+
+def svd_two_site(i, mps, trunc_par=None):
+    r"""Builds a theta and splits it using svd for an MPS.
+
+    Parameters
+    ----------
+    i : int
+        First site.
+    mps : :class:`tenpy.networks.mps.MPS`
+        MPS to use on.
+    trunc_par : None|dict
+       If None no truncation is done. Else dict as in :func:`~tenpy.algorithms.truncation.truncate`.
+    """
+    theta=mps.get_theta(i)
+    theta=theta.combine_legs([['vL','p0'],['p1','vR']], qconj=[+1,-1])
+    if trunc_par==None:
+        u, s, vh, err, renorm = svd_theta(theta, {'chi_max':10000, 'svd_min':1.e-15, 'trunc_cut':1.e-15})
+        mps.norm*=renorm
+    else:
+        u, s, vh, err, renorm = svd_theta(theta, trunc_par)
+        mps.norm*=renorm
+    u=u.split_legs()
+    vh=vh.split_legs()
+    u.ireplace_label('p0', 'p')
+    vh.ireplace_label('p1', 'p')
+    mps.set_B(i, u, form='A')
+    mps.set_B((i+1)%mps.L, vh, form='B')
+    mps.set_SR(i, s)
+
+
+def apply_mpo(psi, mpo, trunc_par):
+    """Applies an mpo and truncates the resulting MPS using SVD.
+
+    Parameters
+    ----------
+    mps : :class:`~tenpy.networks.mps.MPS`
+        MPS to apply operator on
+    mpo : :class:`~tenpy.networks.mpo.MPO`
+        MPO to apply. Usually one of make_UI() or make_UII(). The approximation being made are uncontrolled for other mpos and infinite bc.
+    trunc_par : dict
+        Truncation parameters. See :func:`truncate`
+
+    Returns
+    -------
+    mps : :class:`~tenpy.networks.mps.MPS` 
+        Resulting new MPS
+    """
+    bc=psi.bc
+    if bc!=mpo.bc:
+        raise ValueError("Boundary conditions of MPS and MPO are not the same")
+    if psi.L != mpo.L:
+        raise ValueError("Length of MPS and MPO not the same")
+    Bs=[npc.tensordot(psi.get_B(i, form='B'), mpo.get_W(i), axes=('p', 'p*')) for i in range(psi.L)]
+    Bs[0]=npc.tensordot(psi.get_theta(0,1), mpo.get_W(0), axes=('p0', 'p*'))

```

Comment:
If we simply use `new_mps = mps.MPS(..., form='B',...)`, we should *not* use the `theta` for the first `Bs[0]`.
In other words, delete this line...

Reply:
In the finite case the form is ignored in mps_compress. It's only assumed, that when contracting all tensors in the given form the correct state should be obtained. The leftmost Svalues are part of the state, shouldn't they be included? (In fact it's only cosmetic, as in finite case this is just one, but I think for finite the line is correct.